woocommerce/woocommerce-ios
# Product variations are now editable - description, price, shipping, and inventory

Variation form - part of #2085 

This PR includes various changes and fixes to make a product variation editable on the basic fields - description, price, shipping, and inventory.

## Changes

Please see the differences between a product and variation form in the description of https://github.com/woocommerce/woocommerce-ios/issues/2085, this PR includes all the changes except for the product images!

- Networking layer - a few fixes were made to `ProductVariation` to allow editing inventory and saving a product variation without an image:
  - Updated the logic on how `manageStock` is decoded from the API response: there are two states, whether a variation has its stock management enabled at variation level (`manage_stock` is `true` in the API) or it follows its parent product's stock status (`manage_stock` is `"parent"` in the API). In the past, we've been considered the latter case as the stock management is enabled at variation level but it's not the case. Therefore, the logic and 2 test cases were added on these 2 states, and `manageStock` was updated in a few unit tests that had the incorrect assumption.
  - It now only encodes its `image` field only when the variation has an image. This is due to an [API issue](https://github.com/woocommerce/woocommerce/issues/27116), and if the `image` field isn't set to a non-nil product image the update request will fail (e.g. when updating a variation without an image).
  - `stockStatusKey` is not encoded anymore. Since stock status is only available when a variation's stock management is set to follow its parent product, it's a readonly value from the parent product and shouldn't be encoded.
- App layer:
  - In `ProductVariationsViewController`, navigated to `ProductFormViewController` with a product variation form view model when the user taps on a variation in the list if M3 flag is enabled. Also, `syncingCoordinator.synchronizeFirstPage()` is now only called once in `viewDidLoad` instead of every time in `viewWillAppear`. Updated `ProductDetailsViewModel` to provide `isEditProductsRelease3Enabled` for `ProductVariationsViewController`.
  - In `ProductShippingSettingsViewController`, the shipping class row is hidden for a variation due to an [API issue](https://github.com/woocommerce/woocommerce/issues/27131) (more details in p91TBi-33Q-p2).
  - In `ProductFormViewModelProtocol`, added a new method `func canEditProductSettings() -> Bool` so that the ellipsis menu that shows the product settings in the navigation bar can be hidden for a variation. `ProductFormViewModel` returns `true` for this method, while `ProductVariationFormViewModel` returns `false`. The value of this method is used to determine whether to show the ellipsis menu in the navigation bar in `ProductFormViewController`.
  - In `ProductInventorySettingsViewController`, updated the stock status row to be readonly for a variation: disabled tap action for a variation, and no disclosure indicator is shown in the cell.
  - In `PrimaryFieldRow.name`, added another parameter `isEditable` to be `PrimaryFieldRow.name(name: String?, isEditable: Bool)` because the name field for a variation is readonly. Updated a bunch of related files, and set the value to `product is Product` in `DefaultProductFormTableViewModel`. If the name row is editable, `TextFieldTableViewCell` is used. Otherwise, `BasicTableViewCell` is used with headline style set in `ProductFormTableViewDataSource`.

## Testing

- Go to the Products tab
- Tap on a variable product with non zero variations --> sanity check that the name field is editable
- Tap on the variations row
- Tap on a variation --> a variation form should be presented, without an ellipsis menu in the navigation bar, and the name row is readonly
- Tap on the description row and edit the text
- Go back to the form and tap on the price row, then edit some fields --> no tax section should be shown on price settings
- Go back to the form and tap on the shipping row, then edit some fields --> the shipping class row should not be shown on the shipping settings
- Go back to the form and tap on the inventory row, then toggle the "Manage stock" switch and edit some fields --> the "Limit one per order" toggle row shouldn't be shown as in a product form, and the stock status row is readonly when "Manage stock" is disabled
- Go back to the form and tap "Update" --> the variation should be updated remotely with the changes above

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/89024477-be819400-d357-11ea-9d39-c46a2a3b04b7.gif)

Dark | Light
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-07-31 at 17 59 53](https://user-images.githubusercontent.com/1945542/89024593-eec93280-d357-11ea-8655-1908e75785d3.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-31 at 18 00 31](https://user-images.githubusercontent.com/1945542/89024599-f2f55000-d357-11ea-92ad-33652734068b.png)

Price | Shipping | Inventory (stock management disabled)
-- | -- | --
![Simulator Screen Shot - iPhone 11 - 2020-07-31 at 18 00 00](https://user-images.githubusercontent.com/1945542/89024669-0e605b00-d358-11ea-93ab-7d9e4815b3b2.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-31 at 18 00 08](https://user-images.githubusercontent.com/1945542/89024674-115b4b80-d358-11ea-94aa-cfc27272fa5f.png) | ![Simulator Screen Shot - iPhone 11 - 2020-07-31 at 18 00 12](https://user-images.githubusercontent.com/1945542/89024676-128c7880-d358-11ea-9256-f416fc127aa2.png)


Update release notes:
